File: app/src/main/services/backup-service.ts
import fs from "node:fs/promises";
import path from "node:path";
import Database from "better-sqlite3";

const MAX_QUERY_HISTORY = 20;
const MAX_RESULT_HISTORY = 10;

/** query.sql이 존재하면 workspaceDir/history/query_NNN.sql 로 백업 */
export async function backupQuerySql(workspaceDir: string): Promise<void> {
  const src = path.join(workspaceDir, "query.sql");
  try {
    const content = await fs.readFile(src, "utf-8");
    if (!content.trim()) return;
  } catch {
    return; // query.sql 아직 없음
  }

  const histDir = path.join(workspaceDir, "history");
  await fs.mkdir(histDir, { recursive: true });

  const files = await fs.readdir(histDir).catch(() => [] as string[]);
  const nums = files
    .map((f) => { const m = f.match(/^query_(\d+)\.sql$/); return m ? parseInt(m[1], 10) : null; })
    .filter((n): n is number => n !== null)
    .sort((a, b) => a - b);

  const nextNum = nums.length > 0 ? nums[nums.length - 1] + 1 : 1;
  await fs.copyFile(src, path.join(histDir, `query_${String(nextNum).padStart(3, "0")}.sql`));

  // MAX 초과분 삭제 (오래된 것 먼저)
  if (nums.length >= MAX_QUERY_HISTORY) {
    const toDelete = nums.slice(0, nums.length - MAX_QUERY_HISTORY + 1);
    for (const n of toDelete) {
      await fs.rm(path.join(histDir, `query_${String(n).padStart(3, "0")}.sql`)).catch(() => {});
    }
  }
}

/** DB에 result 테이블이 있으면 result_bak_NNN 으로 복사 */
export function backupResultTable(db: Database.Database): void {
  const exists = db.prepare(
    "SELECT name FROM sqlite_master WHERE type='table' AND name='result'"
  ).get() as { name: string } | undefined;
  if (!exists) return;

  const rows = db.prepare(
    "SELECT name FROM sqlite_master WHERE type='table' AND name LIKE 'result_bak_%'"
  ).all() as { name: string }[];
  const nums = rows
    .map((r) => { const m = r.name.match(/^result_bak_(\d+)$/); return m ? parseInt(m[1], 10) : null; })
    .filter((n): n is number => n !== null)
    .sort((a, b) => a - b);

  const nextNum = nums.length > 0 ? nums[nums.length - 1] + 1 : 1;
  const bakName = `result_bak_${String(nextNum).padStart(3, "0")}`;
  db.exec(`CREATE TABLE "${bakName}" AS SELECT * FROM result`);

  if (nums.length >= MAX_RESULT_HISTORY) {
    const toDelete = nums.slice(0, nums.length - MAX_RESULT_HISTORY + 1);
    for (const n of toDelete) {
      db.exec(`DROP TABLE IF EXISTS "result_bak_${String(n).padStart(3, "0")}"`);
    }
  }
}

/** workspaceDir/history/ 안의 query_NNN.sql 파일 목록을 최신순으로 반환 */
export async function listQueryHistory(workspaceDir: string): Promise<string[]> {
  const histDir = path.join(workspaceDir, "history");
  const files = await fs.readdir(histDir).catch(() => [] as string[]);
  return files
    .filter((f) => /^query_\d+\.sql$/.test(f))
    .sort()
    .reverse();
}

File: server/src/com/ithows/aida/BackupService.java
package com.ithows.aida;

import com.ithows.ResultMap;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.sql.Connection;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 쿼리 히스토리 + result 테이블 백업. 기존 backup-service.ts 포팅.
 */
public final class BackupService {

    private static final int MAX_QUERY_HISTORY = 20;
    private static final int MAX_RESULT_HISTORY = 10;
    private static final Pattern Q = Pattern.compile("^query_(\\d+)\\.sql$");
    private static final Pattern R = Pattern.compile("^result_bak_(\\d+)$");

    private BackupService() {}

    /** query.sql이 있으면 history/query_NNN.sql로 백업(FIFO 최대 20). */
    public static void backupQuerySql(File workspaceDir) {
        try {
            File src = new File(workspaceDir, "query.sql");
            if (!src.exists()) return;
            String content = new String(Files.readAllBytes(src.toPath()), java.nio.charset.StandardCharsets.UTF_8);
            if (content.trim().isEmpty()) return;

            File histDir = new File(workspaceDir, "history");
            histDir.mkdirs();
            List<Integer> nums = nums(histDir.list(), Q);
            int next = nums.isEmpty() ? 1 : nums.get(nums.size() - 1) + 1;
            Files.copy(src.toPath(), new File(histDir, "query_" + pad(next) + ".sql").toPath(),
                    StandardCopyOption.REPLACE_EXISTING);
            if (nums.size() >= MAX_QUERY_HISTORY) {
                for (int i = 0; i <= nums.size() - MAX_QUERY_HISTORY; i++) {
                    new File(histDir, "query_" + pad(nums.get(i)) + ".sql").delete();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /** result 테이블이 있으면 result_bak_NNN으로 복사(FIFO 최대 10). */
    public static void backupResultTable(Connection db) {
        try {
            List<String> tables = SqliteUtil.listTables(db);
            if (!tables.contains("result")) return;
            List<Integer> nums = new ArrayList<>();
            for (String t : tables) {
                Matcher m = R.matcher(t);
                if (m.matches()) nums.add(Integer.parseInt(m.group(1)));
            }
            Collections.sort(nums);
            int next = nums.isEmpty() ? 1 : nums.get(nums.size() - 1) + 1;
            String bak = "result_bak_" + pad(next);
            try (Statement st = db.createStatement()) {
                st.execute("CREATE TABLE " + SqliteUtil.quoteIdent(bak) + " AS SELECT * FROM result");
            }
            if (nums.size() >= MAX_RESULT_HISTORY) {
                for (int i = 0; i <= nums.size() - MAX_RESULT_HISTORY; i++) {
                    try (Statement st = db.createStatement()) {
                        st.execute("DROP TABLE IF EXISTS " + SqliteUtil.quoteIdent("result_bak_" + pad(nums.get(i))));
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /** history/query_NNN.sql 목록을 최신순으로 반환. */
    public static List<String> listQueryHistory(File workspaceDir) {
        File histDir = new File(workspaceDir, "history");
        String[] files = histDir.list();
        List<String> out = new ArrayList<>();
        if (files != null) {
            for (String f : files) if (Q.matcher(f).matches()) out.add(f);
        }
        Collections.sort(out);
        Collections.reverse(out);
        return out;
    }

    private static List<Integer> nums(String[] files, Pattern p) {
        List<Integer> nums = new ArrayList<>();
        if (files != null) {
            for (String f : files) {
                Matcher m = p.matcher(f);
                if (m.matches()) nums.add(Integer.parseInt(m.group(1)));
            }
        }
        Collections.sort(nums);
        return nums;
    }

    private static String pad(int n) { return String.format("%03d", n); }
}
